# MySQL in Jupyter

This is an example notebook to illustrate how to use MySQL from within a Jupyter notenook.  
We will walk through connecting to a MySQL server, how to make form simple SQL queries,  
how to capture large result sets, and how to download the results in CSV format (to make  
it easy to get subsets of a database into an Excel spreadsheet in your PC.

> "Everything should be made as simple as possible, but not simpler."  
 \- Albert Einstein

***

## MySQL example tutorial

In this example you will learn how to connect to the database server,  
see which databases are installed, discover the structure of tables in a database.

Once we understand the database structure, we can run queries against the tables in  
the database, and then save the results for further analysis.

***

### step 1: load the sql library

Before we do anything else, we must first tell Jupyter to load a sql 'magic' module.  
The sql module enables us to make database connections and query the database.

The command to load the sql 'magic' is
```python
load_ext sql
```


In [1]:
load_ext sql

***

### step 2: connect to the database

After the sql module is loaded, we tell it to log into a MySQL database.

The command 
```mySQL
mysql://tester:badpassword@mysqlserver/employees
```
logs us into the mysql server at the server 'mysqlserver' as the user 'tester' and selects
the database named 'employees'



In [2]:
%sql mysql://tester:badpassword@mysqlserver/employees

'Connected: tester@employees'

***
### step 3: SQL commands to find databases and tables

Let's see what databases are available on this server.

The command
```mySQL
show databases
```
will list all the databases available to us. MySQL usues some database  
internally to keep track of settings and other housekeeping tasks, so  
here we will see a housekeeping databases (information_schema). 

For this course, we are only interested in the database named 'employees'.

In [3]:
%sql show databases;

4 rows affected.


[('information_schema',), ('employees',), ('mysql',), ('performance_schema',)]

The information_schema database is where mysql does some bookkeeping on  
the structure of other databases. The 'employees' database contains the  
tables we want to look at since it has employee, department, and salary data.

If we don't already know the structure of the employees database, we can  
ask MySQL to tell us about the tables that make up the database with the command
```mySQL
show tables;
```


In [4]:
%sql show tables;

6 rows affected.


[('departments',),
 ('dept_emp',),
 ('dept_manager',),
 ('employees',),
 ('salaries',),
 ('titles',)]

You can see that there are 6 tables in this database. 
***

## step 4: Discover the table structure

How do can we discover the table structures for each of these tables? 

If we know the names of a table, we can ask mySQL to describe it.  
For example, we can ask for a description of the departments table with the command
```mySQL
describe departments;
```



In [5]:
%sql describe departments;

2 rows affected.


[('dept_no', 'char(4)', 'NO', 'PRI', None, ''),
 ('dept_name', 'varchar(40)', 'NO', 'UNI', None, '')]

***
## step 5: How big is the table?

Its a good idea to have a rough sense of how big a table is before  
you start running queries so you don't accidentally request a couple  
million rows of results. 

To do this we want a query that will count all the rows. The command to do this is
```mySQL
select count(*) from departments;
```
Let's see how big the departments table is:

In [6]:
%sql select count(*) from departments;

1 rows affected.


[(9,)]

The departments table is not very big, so it should be reasonably safe  
to query for everything in the table. 

The command
```mySQL
select * from departments;
```
will match every column in every row in the table. 

For this table we get 9 rows of results:

In [7]:
%sql select * from departments;

9 rows affected.


[('d009', 'Customer Service'),
 ('d005', 'Development'),
 ('d002', 'Finance'),
 ('d003', 'Human Resources'),
 ('d001', 'Marketing'),
 ('d004', 'Production'),
 ('d006', 'Quality Management'),
 ('d008', 'Research'),
 ('d007', 'Sales')]

***

## step 6: Selecting a subset of a table

The employees table is much larger (and more complex) than the departments table.

Let's use commands we already know to discover the table structure and size:

In [8]:
%sql describe employees

6 rows affected.


[('emp_no', 'int(11)', 'NO', 'PRI', None, ''),
 ('birth_date', 'date', 'NO', '', None, ''),
 ('first_name', 'varchar(14)', 'NO', '', None, ''),
 ('last_name', 'varchar(16)', 'NO', '', None, ''),
 ('gender', "enum('M','F')", 'NO', '', None, ''),
 ('hire_date', 'date', 'NO', '', None, '')]

In [9]:
%sql select count(*) from employees

1 rows affected.


[(300024,)]

We probably don't want to select all 300,024 rows in the employees table,  
so we are going to need a better selet statement. 

Select statements can include an optional 'where' clause to limit the scope  
of the select. By limiting scope we can cut down on the number of rows returned.

Since we know the names of the columns in the table, we could limit the select  
scope to the rows that have a specific value for one of the columns. 

For example, if I wanted to count all the of employees with the first name 'Mark' 
I could add the clause
```mySQL
where first_name = 'Mark'
```
to the select statement I previously used to count the rows in the table. 

So, to count all the employees named Mark, my select satatement would be
```mySQL
select count(*) from employees where first_name = 'Mark'
```


In [10]:
%sql select count(*) from employees where first_name = 'Mark'

1 rows affected.


[(230,)]

230 rows of results looks like a managable number, so let's get the detailed results with this query
```SQL
select * from employees where first_name = 'Mark'
```

In [11]:
%sql select * from employees where first_name = 'Mark'

230 rows affected.


[(10415, datetime.date(1957, 11, 12), 'Mark', 'Coorg', 'M', datetime.date(1993, 10, 25)),
 (13517, datetime.date(1961, 2, 22), 'Mark', 'Setlzner', 'M', datetime.date(1988, 1, 2)),
 (14248, datetime.date(1955, 10, 13), 'Mark', 'Alblas', 'M', datetime.date(1989, 5, 24)),
 (14335, datetime.date(1955, 9, 16), 'Mark', 'Fraisse', 'M', datetime.date(1997, 10, 9)),
 (15031, datetime.date(1956, 12, 22), 'Mark', 'Naumovich', 'M', datetime.date(1992, 9, 7)),
 (17371, datetime.date(1958, 2, 3), 'Mark', 'Sankaranarayanan', 'F', datetime.date(1988, 10, 16)),
 (17551, datetime.date(1954, 2, 21), 'Mark', 'Foote', 'F', datetime.date(1998, 10, 4)),
 (19561, datetime.date(1957, 2, 23), 'Mark', 'Kinley', 'F', datetime.date(1987, 4, 30)),
 (20189, datetime.date(1955, 2, 19), 'Mark', 'Nanard', 'M', datetime.date(1986, 8, 28)),
 (20945, datetime.date(1955, 12, 22), 'Mark', 'Marsiglia', 'F', datetime.date(1998, 7, 3)),
 (22481, datetime.date(1953, 7, 13), 'Mark', 'Ciolek', 'F', datetime.date(1991, 4, 9)),
 (23195, datetime.date(1952, 5, 27), 'Mark', 'Ertl', 'M', datetime.date(1986, 9, 25)),
 (24632, datetime.date(1958, 11, 3), 'Mark', 'Harllee', 'M', datetime.date(1989, 9, 22)),
 (27746, datetime.date(1964, 11, 29), 'Mark', 'Lunn', 'M', datetime.date(1987, 9, 22)),
 (31568, datetime.date(1965, 1, 10), 'Mark', 'Lalonde', 'M', datetime.date(1985, 12, 24)),
 (33764, datetime.date(1956, 10, 9), 'Mark', 'Deville', 'M', datetime.date(1985, 4, 29)),
 (37408, datetime.date(1955, 10, 28), 'Mark', 'Boguraev', 'M', datetime.date(1990, 9, 16)),
 (38410, datetime.date(1958, 6, 19), 'Mark', 'Melichar', 'M', datetime.date(1992, 2, 17)),
 (39969, datetime.date(1955, 10, 15), 'Mark', 'Conia', 'M', datetime.date(1985, 12, 9)),
 (42974, datetime.date(1962, 7, 27), 'Mark', 'Brookman', 'M', datetime.date(1989, 12, 6)),
 (43153, datetime.date(1963, 8, 27), 'Mark', 'McConalogue', 'F', datetime.date(1997, 8, 11)),
 (44900, datetime.date(1965, 1, 5), 'Mark', 'Schwaller', 'F', datetime.date(1989, 9, 6)),
 (46720, datetime.date(1964, 7, 4), 'Mark', 'Sethi', 'M', datetime.date(1986, 3, 27)),
 (47842, datetime.date(1960, 8, 16), 'Mark', 'Sigstam', 'M', datetime.date(1989, 5, 21)),
 (50359, datetime.date(1963, 9, 23), 'Mark', 'Feldmann', 'M', datetime.date(1987, 3, 25)),
 (50872, datetime.date(1955, 1, 12), 'Mark', 'Hanabata', 'F', datetime.date(1993, 10, 31)),
 (51175, datetime.date(1953, 8, 18), 'Mark', 'Lorho', 'M', datetime.date(1987, 4, 13)),
 (52014, datetime.date(1962, 7, 5), 'Mark', 'Peres', 'M', datetime.date(1994, 9, 23)),
 (54842, datetime.date(1961, 2, 23), 'Mark', 'Nitto', 'M', datetime.date(1994, 12, 21)),
 (56628, datetime.date(1955, 10, 21), 'Mark', 'Esposito', 'F', datetime.date(1986, 6, 28)),
 (56767, datetime.date(1952, 6, 27), 'Mark', 'Trelles', 'M', datetime.date(1988, 5, 6)),
 (57141, datetime.date(1960, 3, 28), 'Mark', 'Oxenboll', 'M', datetime.date(1986, 1, 28)),
 (57724, datetime.date(1955, 10, 24), 'Mark', 'Cools', 'F', datetime.date(1994, 2, 20)),
 (58209, datetime.date(1957, 11, 17), 'Mark', 'Ermel', 'M', datetime.date(1989, 2, 19)),
 (58367, datetime.date(1964, 9, 15), 'Mark', 'Goldhammer', 'F', datetime.date(1993, 9, 29)),
 (59357, datetime.date(1964, 7, 6), 'Mark', 'Barriga', 'M', datetime.date(1986, 12, 5)),
 (61274, datetime.date(1955, 1, 6), 'Mark', 'Antonisse', 'M', datetime.date(1997, 3, 18)),
 (62284, datetime.date(1961, 10, 10), 'Mark', 'Hebden', 'M', datetime.date(1987, 2, 5)),
 (63343, datetime.date(1953, 4, 22), 'Mark', 'Ananiadou', 'M', datetime.date(1988, 9, 28)),
 (63478, datetime.date(1957, 11, 4), 'Mark', 'Godskesen', 'F', datetime.date(1986, 3, 26)),
 (64178, datetime.date(1959, 3, 14), 'Mark', 'Conde', 'F', datetime.date(1986, 1, 11)),
 (64961, datetime.date(1960, 8, 23), 'Mark', 'Jording', 'M', datetime.date(1986, 6, 29)),
 (65161, datetime.date(1957, 1, 3), 'Mark', 'Dymetman', 'F', datetime.date(1987, 7, 24)),
 (65568, datetime.date(1952, 8, 9), 'Mark', 'Ventosa', 'M', datetime.date(1988, 11, 26)),
 (67026, datetime.date(1960, 6, 

***

## step 7: Saving query results to a file

We have a query that returns a subset of the employees table,  
but how can you get this onto your computer in a form that  
Excel will understand?

You could use your mouse to copy the results displayed on this page,  
open Excel and then paste the results into a spreadsheet. That approach  
works well enough for small data sets, but what about really large data?

The best way to handle big datasets is to save the results to a file,  
and then download that file to your PC.

The first step is to run the query again, and tell sql to put the results  
in a variable.

For instance, if we wanted to put the query results from above into a variable  
named "resultset1" can can issue this command:
```mySQL
resultset1 = %sql select * from employees where first_name = 'Mark'
```



In [12]:
resultset1 = %sql select * from employees where first_name = 'Mark'

230 rows affected.


Instead of displaying all the rows that were returned by the query, the notebook simply  
displayed the number of rows returned. This is very convenient for handling large amounts of data.

Now that we have a variable with the data, we can use a little python code to reformat  
the results into CSV (Comma Separated Value) format, which is something Excel understands.

We can do this by importing the csv package (which knows how to write CSV formatted files).
```python
import csv
```

we then open a file named 'resultset1.csv', and make a filewriter that reads each row in  
the viable that holds the results and writes it out in CSV format to the file
```python
with open('resultset1.csv', 'w', newline='') as csvfile:
    filewriter = csv.writer(csvfile, dialect='excel')
    for row in resultset1:
        filewriter.writerow(row)
```

We can combine these two steps to create a CSV formatted output file:

In [13]:
import csv
with open('resultset1.csv', 'w', newline='') as csvfile:
    filewriter = csv.writer(csvfile, dialect='excel')
    for row in resultset1:
        filewriter.writerow(row)

***
To verify that the output file exists, let's ask the notebook to display the current directory using the command
```bash
ls -l
```

The directory listing from the 'ls' command should include a file named
```
resultset1.csv
```


In [14]:
ls -l


total 248
-rw-r--r-- 1 jovyan  1000 237091 Sep 28 18:22 mysql-example.ipynb
-rw-rw-r-- 1 jovyan  1000   1218 Sep 28 18:15 restore-my-notebook.ipynb
-rw-rw-r-- 1 jovyan users  10307 Sep 28 18:23 resultset1.csv


***
## step 8: Download the results file

At the top of this page, underneath the 'Jupyter' name and icon, there are several options  
(File, Edit, View, Insert, etc...). You can download any file that you are viewing by  
selecting the download option under the File menu (once you are viewing the file).

In other words, to download the CSV results file we just created, we need to open it,  
then select the download option from the File menu in the window with the results after  
we open them.

There are several ways to open the resultset1.csv file. You can
- [click this link](/edit/resultset1.csv "open resultset1.csv") to open resultset1.csv
- select resultset1.csv from the "Open" option under the File menu
- click on the Jupyter icon at the top of this page to get a list of files, then select the resultset1.csv file

***
# Advanced topics

## How many different first names are there?

The clause 'distinct' tells SQL to return a collection of unique values rather than
every value.  
This is useful if you want to find out how many different first names 
there are in the employees table.

The command 
```SQL
select distinct first_name from employees;
```
will give us a list of the unique first names, but the list is not necessarily in any particular order.  
We can add an optional 'order by' clause to that command to tell SQL how to order the results. 

For example, if we want to order the results in reverse alphabetical order, the clause
```SQL
order by first_name desc
```
tells mySQL which column to order by and to order in descending (desc) order.

Putting this all together we get a command to get unique names in reverse alphbetical order:
```SQL
select distinct first_name from employees order by first_name desc;
```

In [15]:
%sql select distinct first_name from employees order by first_name desc;

1275 rows affected.


[('Zvonko',),
 ('Zsolt',),
 ('Zorica',),
 ('Zongyan',),
 ('Ziyad',),
 ('Ziya',),
 ('Ziva',),
 ('Zito',),
 ('Zita',),
 ('Zissis',),
 ('Zine',),
 ('Zijian',),
 ('Zhongwei',),
 ('Zhonghui',),
 ('Zhonghua',),
 ('Zhiwei',),
 ('Zhilian',),
 ('Zhiguo',),
 ('Zhigen',),
 ('Zhensheng',),
 ('Zhenhua',),
 ('Zhenbing',),
 ('Zhaofang',),
 ('Zhanqiu',),
 ('Zeydy',),
 ('Zengping',),
 ('Zeljko',),
 ('Zejun',),
 ('Zdislav',),
 ('Yuuichi',),
 ('Yuriy',),
 ('Yurij',),
 ('Yurii',),
 ('Yuqun',),
 ('Yuping',),
 ('Yunming',),
 ('Yuichiro',),
 ('Yuguang',),
 ('Yuchang',),
 ('Yucel',),
 ('Yucai',),
 ('Youssef',),
 ('Youpyo',),
 ('Younwoo',),
 ('Youngkon',),
 ('Youjian',),
 ('Yoshinari',),
 ('Yoshimitsu',),
 ('Yongqiao',),
 ('Yongmin',),
 ('Yongmao',),
 ('Yonghong',),
 ('Yonghoan',),
 ('Yongdong',),
 ('Yolla',),
 ('Ymte',),
 ('Yishay',),
 ('Yishai',),
 ('Yinlin',),
 ('Yinghua',),
 ('Yifei',),
 ('Yefim',),
 ('Yechiam',),
 ('Yechezkel',),
 ('Yannis',),
 ('Yannik',),
 ('Yakkov',),
 ('Yakichi',),
 ('Yahiko',),
 ('Yagil',),
 ('Yaghout',),
 ('Xumin',),
 ('Xuejun',),
 ('Xuejia',),
 ('Xuedong',),
 ('Xudong',),
 ('Xiping',),
 ('Xiong',),
 ('Xinyu',),
 ('Xinglin',),
 ('Xinan',),
 ('Xiaoshan',),
 ('Xiaoqiu',),
 ('Xiaoqiang',),
 ('Xiaopeng',),
 ('Xiaoheng',),
 ('Xiaocheng',),
 ('Xiaobin',),
 ('Xianlong',),
 ('Xiahua',),
 ('Xiadong',),
 ('Wuxu',),
 ('Woody',),
 ('Wonhee',),
 ('Wojceich',),
 ('Wilmer',),
 ('Willard',),
 ('Werner',),
 ('Weiyi',),
 ('Weiye',),
 ('Weiwu',),
 ('Weiru',),
 ('Weijing',),
 ('Weidon',),
 ('Weicheng',),
 ('Wayne',),
 ('Wanqing',),
 ('Waiman',),
 ('Vugranam',),
 ('Volkmar',),
 ('Vojin',),
 ('Vivian',),
 ('Vitaly',),
 ('Vitali',),
 ('Visit',),
 ('Vishwani',),
 ('Vishv',),
 ('Vishu',),
 ('Vincent',),
 ('Vidya',),
 ('Vidar',),
 ('Vesna',),
 ('Venkatesan',),
 ('Vatsa',),
 ('Vasiliy',),
 ('Vasilis',),
 ('Vasilii',),
 ('Van',),
 ('Valter',),
 ('Valeska',),
 ('Valery',),
 ('Valeri',),
 ('Valdiodio',),
 ('Val',),
 ('Uzi',),
 ('Uwe',),
 ('Utz',),
 ('Utpal',),
 ('Ute',),
 ('Uta',),
 ('Urs',),
 ('Uriel',),
 ('Uri',),
 ('Uno',),
 ('Ung',),
 ('Unal',),
 ('Uma',),
 ('Ult',),
 ('Ulises',),
 ('Uli',),
 ('Ulf',),
 ('Uinam',),
 ('Ugo',),
 ('Udo',),
 ('Udi',),
 ('Udaiprakash',),
 ('Tzvetan',),
 ('Tzu',),
 ('Tze',),
 ('Tuval',),
 ('Turgut',),
 ('Tua',),
 ('Tsz',),
 ('Tsvetan',),
 ('Tsutomu',),
 ('Tsuneo',),
 ('Tse',),
 ('Trygve',),
 ('True',),
 ('Tru',),
 ('Troy',),
 ('Tristan',),
 ('Tremaine',),
 ('Tran',),
 ('Toshimori',),
 ('Toshimo',),
 ('Toshimitsu',),
 ('Toshimi',),
 ('Toshiko',),
 ('Toshiki',),
 ('Torsten',),
 ('Tooru',),
 ('Toong',),
 ('Tonny',),
 ('Tonia',),
 ('Toney',),
 ('Tomoyuki',),
 ('Tomokazu',),
 ('Tomofumi',),
 ('Tommaso',),
 ('Tokuyasu',),
 ('Tiina',),
 ('Tianruo',),
 ('Thodoros',),
 ('Theron',),
 ('Theirry',),
 ('Thanasis',),
 ('Teunis',),
 ('Tetsushi',),
 ('Tetsurou',),
 ('Tesuya',),
 ('Tesuro',),
 ('Teruyuki',),
 ('Terresa',),
 ('Temple',),
 ('Teiji',),
 ('Tayeb',),
 ('Tamiya',),
 ('Takushi',),
 ('Takanari',),
 ('Takahito',),
 ('Takahiro',),
 ('Takahira',),
 ('Taizo',),
 ('Taisook',),
 ('Taiji',),
 ('Taegyun',),
 ('Tadanori',),
 ('Tadahiro',),
 ('Tadahiko',),
 ('Szabolcs',),
 ('Syozo',),
 ('Sverrir',),
 ('Sven',),
 ('Suzette',),
 ('Suvo',),
 ('Susumu',),
 ('Susuma',),
 ('Sushant',),
 ('Susanta',),
 ('Susanne',),
 ('Susanna',),
 ('Surveyors',),
 ('Supot',),
 ('Sungwon',),
 ('Sumant',),
 ('Sumali',),
 ('Sukumar',),
 ('Sugwoo',),
 ('Suebskul',),
 ('Sudharsan',),
 ('Sudhanshu',),
 ('Subir',),
 ('Subhrajyoti',),
 ('Subhash',),
 ('Subhankar',),
 ('Subhada',),
 ('Subbu',),
 ('Stevo',),
 ('Steve',),
 ('Stepehn',),
 ('Stella',),
 ('Stein',),
 ('Steen',),
 ('Stamatina',),
 ('Stafford',),
 ('Stabislas',),
 ('Srinidhi',),
 ('Sreenivas',),
 ('Sreekrishna',),
 ('Spyrose',),
 ('Souichi',),
 ('Sorina',),
 ('Somnath',),
 ('Soenke',),
 ('Snehasis',),
 ('Slavian',),
 ('Sivanarayana',),
 ('Sigeru',),
 ('Siddarth',),
 ('Sibyl',),
 ('Siamak',),
 ('Shuzo',),
 ('Shushma',),
 ('Shunichi',),
 ('Shuky',),
 ('Shuji',),
 ('Shuichi',),
 ('Shuho',),
 ('Shugo',),
 ('Shrikan

***
## Test your understanding

Now that you have worked through the tutorial it is time to see well do you understand the material.  
Can you answer the questions below by creating your own SQL queries?
***

### 1.) How many different last names are there in the employees table?

***
### 2.) How many different salaries are there?